core/src/main/java/org/radargun/config/ClasspathScanner.java

```diff
@@ -0,0 +1,103 @@
+
+import org.radargun.logging.Log;
+import org.radargun.logging.LogFactory;
+
+/**
+ * Helper for listing classes on classpath.
+ *
+ * @author Radim Vansa &lt;rvansa@redhat.com&gt;
+ */
+public class ClasspathScanner {
+
+   private static Log log = LogFactory.getLog(ClasspathScanner.class);
+   private static final PrintStream NULL_PRINT_STREAM = new PrintStream(new OutputStream() {
+      @Override
+      public void write(int b) throws IOException {}
+   });
+   private static final PrintStream ERR_PRINT_STREAM = System.err;
+
+   public static <TClass, TAnnotation extends Annotation> void scanClasspath(
+         Class<TClass> superClass, Class<TAnnotation> annotationClass, String requirePackage, Consumer<Class<? extends TClass>> consumer) {
+      String classPath = System.getProperty("java.class.path");
+      String[] classPathParts = classPath.split(File.pathSeparator);
+
+      for (String resource : classPathParts) {
+         File resourceFile = new File(resource);
+         if (resourceFile.isFile()) {
+            scanFile(resource, 0, superClass, annotationClass, requirePackage, consumer);
+         } else if (resourceFile.isDirectory()) {
+            int prefixLength = resource.toString().length() + 1;
+            try {
+               Files.find(resourceFile.toPath(), Integer.MAX_VALUE, (path, attrs) -> attrs.isRegularFile() && (path.toString().endsWith(".class") || path.toString().endsWith(".jar"))).forEach(file -> {
+                  scanFile(file.toString(), prefixLength, superClass, annotationClass, requirePackage, consumer);
+               });
+            } catch (IOException e) {
+               throw new RuntimeException(e);
+            }
+         }
+      }
+   }
+
+   protected static <TClass, TAnnotation extends Annotation> void scanFile(String resource, int prefixLength,
+         Class<TClass> superClass, Class<TAnnotation> annotationClass, String requirePackage, Consumer<Class<? extends TClass>> consumer) {
+      if (resource.endsWith(".jar")) {
+         scanJar(resource, superClass, annotationClass, requirePackage, consumer);
+      } else if (resource.endsWith(".class")) {
+         String className = resource.substring(prefixLength, resource.length() - 6).replaceAll("/", ".");
+         scanClass(className, superClass, annotationClass, requirePackage, consumer);
+      }
+   }
+
+   public static <TClass, TAnnotation extends Annotation> void scanJar(String path,
+         Class<TClass> superClass, Class<TAnnotation> annotationClass, String requirePackage, Consumer<Class<? extends TClass>> consumer) {
+      log.tracef("Looking for @%s %s, loading classes from %s", annotationClass.getSimpleName(), superClass.getSimpleName(), path);
+      try (ZipInputStream inputStream = new ZipInputStream(new FileInputStream(path))) {
+         for(;;) {
+            ZipEntry entry = inputStream.getNextEntry();
+            if (entry == null) break;
+            if (!entry.getName().endsWith(".class")) continue;
```

Comment:
I think introducing .class constant and using entry.getName().length() - CLASS.length would be more obvious at first sight.